mithril-aggregator/src/artifact_builder/cardano_database_artifacts/immutable.rs

```diff
@@ -0,0 +1,788 @@
+use slog::{error, Logger};
+
+use mithril_common::{
+    digesters::IMMUTABLE_DIR,
+    entities::{CompressionAlgorithm, ImmutableFileNumber, ImmutablesLocation, MultiFilesUri},
+    logging::LoggerExtensions,
+    StdResult,
+};
+
+use crate::{file_uploaders::LocalUploader, FileUploader, Snapshotter};
+
+fn default_extractor(file_uri: &str) -> StdResult<Option<String>> {
+    let regex = Regex::new(r".*(\d{5})")?;
+
+    Ok(regex
+        .captures(file_uri)
+        .and_then(|mat| mat.get(1))
+        .map(|immutable_match| {
+            let mut template = file_uri.to_string();
+            template.replace_range(immutable_match.range(), "{immutable_file_number}");
+
+            template
+        }))
+}
+
+/// The [ImmutableFilesUploader] trait allows identifying uploaders that return locations for immutable files archive.
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub trait ImmutableFilesUploader: Send + Sync {
+    /// Uploads the archives at the given filepaths and returns the location of the uploaded file.
+    async fn batch_upload<'a>(&self, filepaths: &[&'a Path]) -> StdResult<ImmutablesLocation>;
+}
+
+#[async_trait]
+impl ImmutableFilesUploader for LocalUploader {
+    async fn batch_upload<'a>(&self, filepaths: &[&'a Path]) -> StdResult<ImmutablesLocation> {
+        let mut file_uris = Vec::new();
+        for filepath in filepaths {
+            file_uris.push(self.upload(filepath).await?.into());
+        }
+
+        let template_uri = MultiFilesUri::extract_template_from_uris(file_uris, default_extractor)?
+            .ok_or_else(|| anyhow!("No template found in the uploaded files"))?;
+
+        Ok(ImmutablesLocation::CloudStorage {
+            uri: MultiFilesUri::Template(template_uri),
+        })
+    }
+}
+
+pub struct ImmutableArtifactBuilder {
+    uploaders: Vec<Arc<dyn ImmutableFilesUploader>>,
+    snapshotter: Arc<dyn Snapshotter>,
+    compression_algorithm: CompressionAlgorithm,
+    logger: Logger,
+}
+
+impl ImmutableArtifactBuilder {
+    pub fn new(
+        uploaders: Vec<Arc<dyn ImmutableFilesUploader>>,
+        snapshotter: Arc<dyn Snapshotter>,
+        compression_algorithm: CompressionAlgorithm,
+        logger: Logger,
+    ) -> StdResult<Self> {
+        if uploaders.is_empty() {
+            return Err(anyhow!(
+                "At least one uploader is required to create an 'ImmutableArtifactBuilder'"
+            ));
+        }
+
+        Ok(Self {
+            uploaders,
+            snapshotter,
+            compression_algorithm,
+            logger: logger.new_with_component_name::<Self>(),
+        })
+    }
+
+    pub async fn upload(
+        &self,
+        up_to_immutable_file_number: ImmutableFileNumber,
+    ) -> StdResult<Vec<ImmutablesLocation>> {
+        let archives_paths =
+            self.immutable_archives_paths_creating_the_missing_ones(up_to_immutable_file_number)?;
+        let archives_paths: Vec<_> = archives_paths.iter().map(Path::new).collect();
```

Comment:
Maybe this conversion could be made implicit?

Reply:
I can do it with a generic on `upload_immutable_archives`.
```
async fn upload_immutable_archives<P: AsRef<Path>>(
        &self,
        archive_paths: &[P],
    ) -> StdResult<Vec<ImmutablesLocation>> {
```
On that case, I need to make this conversion in  `upload_immutable_archives` before calling `batch_upload`. 
To avoid that, we can modify the `ImmutableFilesUploader` trait and use a generic AsRef<Path>. 
Do you think it's could be a good thing ?